README.md
# sensu-wavefront-handler

## Table of Contents
- [Overview](#overview)
- [Usage Examples](#usage-examples)
- [Configuration](#configuration)
  - [Asset registration](#asset-registration)
  - [Resource definition](#resource-definition)
- [Installation from source](#installation-from-source)
- [Contributing](#contributing)

## Overview
The Sensu Wavefront Handler is a [Sensu Event Handler][9] that sends metrics to the SaaS time-series
database [Wavefront][10] via a [proxy][12]. [Sensu][11] can collect metrics using check output
metric extraction or the StatsD listener. Those collected metrics pass through the event pipeline,
allowing Sensu to deliver normalized metrics to the configured metric event handlers. This Wavefront
handler will allow you to store, instrument, and visualize the metric data from Sensu.

## Usage Examples

Help:
```
sends metrics to a wavefront proxy using the wavefront data format

Usage:
  sensu-wavefront-handler [flags]
  sensu-wavefront-handler [command]

Available Commands:
  help        Help about any command
  version     Print the version number of this plugin

Flags:
  -f, --flush-interval-seconds int   the flush interval of the wavefront proxy (in seconds) (default 1)
  -h, --help                         help for sensu-wavefront-handler
      --host string                  the host of the wavefront proxy (default "127.0.0.1")
  -m, --metrics-port int             the port of the wavefront proxy (default 2878)
  -p, --prefix string                the string to be prepended to the metric name
  -t, --tags stringToString          the additional tags to merge with the metric tags (default [])
```

## Configuration

### Asset registration

[Sensu Assets][14] are the best way to make use of this plugin. If you're not using an asset, please
consider doing so! If you're using sensuctl 5.13 with Sensu Backend 5.13 or later, you can use the
following command to add the asset:

```
sensuctl asset add sensu/sensu-wavefront-handler
```

If you're using an earlier version of sensuctl, you can find the asset on the [Bonsai Asset Index][13].

### Resource definition

```yml
---
type: Handler
api_version: core/v2
metadata:
  name: sensu-wavefront-handler
  namespace: default
spec:
  command: sensu-wavefront-handler --host "127.0.0.1" --metrics-port 2878 --prefix sensu --tags type="system" --flush-interval-seconds 1
  runtime_assets:
  - sensu-wavefront-handler
  type: pipe
```

## Installation from source

The preferred way of installing and deploying this plugin is to use it as an Asset. If you would
like to compile and install the plugin from source or contribute to it, download the latest version
or create an executable script from this source.

From the local path of the sensu-wavefront-handler repository:

```
go build
```

## Contributing

For more information about contributing to this plugin, see [Contributing][1].

[1]: https://github.com/sensu/sensu-go/blob/master/CONTRIBUTING.md
[2]: github.com/sensu-community/sensu-plugin-sdk
[3]: https://github.com/sensu-plugins/community/blob/master/PLUGIN_STYLEGUIDE.md
[4]: https://github.com/sensu/sensu-wavefront-handler/blob/master/.github/workflows/release.yml
[5]: https://github.com/sensu/sensu-wavefront-handler/actions
[6]: https://github.com/sensu/sensu-wavefront-handler/releases
[7]: https://github.com/sensu/sensu-wavefront-handler/blob/master/main.go
[8]: https://bonsai.sensu.io/
[9]: https://docs.sensu.io/sensu-go/latest/reference/handlers/#how-do-sensu-handlers-work
[10]: https://www.wavefront.com/
[11]: https://github.com/sensu/sensu-go
[12]: https://docs.wavefront.com/proxies.html
[13]: https://bonsai.sensu.io/assets/sensu/sensu-wavefront-handler
[14]: https://docs.sensu.io/sensu-go/latest/reference/assets/


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased
### Changed
- Updated to latest plugin sdk version v0.16.0  for build dependancy
- Added bonsai webhook notification github action workflow for release workflow

## [0.3.0] - 2022-04-22

### Changed
- Updated to latest wavefront sdk patch release 0.9.11 for build dependancy
- Updated to latest plugin sdk version v0.16.0-alpha4  for build dependancy
- Updated to require go 1.18 for build dependancy

### Fixed
- Coerce metric timestamp sent to wavefront to 1 second precision using heuristic to determine Sensu metric timestamp precision. Needed to correctly handle prometheus metrics with millisecond precision ingested by agents.

## [0.2.2] - 2022-03-14

### Changed
- Updated go version to 1.17

## [0.2.1] - 2022-03-14

### Changed
- Updated release platform targets

## [0.2.0] - 2022-03-14

### Changed
- Update Sensu Go and SDK dependencies with the correct modules

## [0.1.0] - 2020-01-07

### Added
- Initial release


main_test.go
package main

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"net"
	"strconv"
	"strings"
	"testing"

	"github.com/google/go-cmp/cmp"
	corev2 "github.com/sensu/sensu-go/api/core/v2"
)

func TestExecuteHandler(t *testing.T) {
	event := corev2.FixtureEvent("entity1", "check1")
	event.Check = nil
	event.Metrics = corev2.FixtureMetrics()
	event.Metrics.Points[0].Timestamp = 1580922166749062000

	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	defer ln.Close()
	errC := make(chan error, 1)
	go func() {
		defer close(errC)
		conn, err := ln.Accept()
		if err != nil {
			errC <- err
			return
		}
		var buf bytes.Buffer
		if _, err := io.Copy(&buf, conn); err != nil {
			errC <- err
			return
		}
		want := `"answer" 42 1580922166 source="entity1" "foo"="bar"`
		if got := strings.TrimSpace(buf.String()); got != want {
			errC <- errors.New(cmp.Diff(got, want))
		}
	}()

	host, portS, err := net.SplitHostPort(ln.Addr().String())
	if err != nil {
		t.Fatal(err)
	}
	handlerConfig.Host = host
	port, err := strconv.Atoi(portS)
	if err != nil {
		t.Fatal(err)
	}
	handlerConfig.MetricsPort = port
	if err := executeHandler(event); err != nil {
		t.Fatal(err)
	}
	if err := <-errC; err != nil {
		t.Error(err)
	}
}

func TestSecTimestamp(t *testing.T) {
	tests := []struct {
		In   int64
		Want int64
	}{
		{
			In:   1580922166749062000,
			Want: 1580922166,
		},
		{
			In:   1580922166749062,
			Want: 1580922166,
		},
		{
			In:   1580922166749,
			Want: 1580922166,
		},
		{
			In:   1580922166,
			Want: 1580922166,
		},
	}
	for _, test := range tests {
		t.Run(fmt.Sprintf("%d => %d", test.In, test.Want), func(t *testing.T) {
			if got, want := secTimestamp(test.In), test.Want; got != want {
				t.Error(cmp.Diff(got, want))
			}
		})
	}
}


main.go
package main

import (
	"fmt"
	"log"
	"math"

	corev2 "github.com/sensu/sensu-go/api/core/v2"
	"github.com/sensu/sensu-plugin-sdk/sensu"
	wavefront "github.com/wavefronthq/wavefront-sdk-go/senders"
)

// Config represents a handler config for the wavefront handler.
type Config struct {
	sensu.PluginConfig
	Host                 string
	MetricsPort          int
	FlushIntervalSeconds int
	Prefix               string
	Tags                 map[string]string
}

const (
	host   = "host"
	port   = "metrics-port"
	flush  = "flush-interval-seconds"
	prefix = "prefix"
	tags   = "tags"
)

var (
	handlerConfig = Config{
		PluginConfig: sensu.PluginConfig{
			Name:     "sensu-wavefront-handler",
			Short:    "sends metrics to a wavefront proxy using the wavefront data format",
			Keyspace: "sensu.io/plugins/sensu-wavefront-handler/config",
		},
	}

	opts = []sensu.ConfigOption{
		&sensu.PluginConfigOption[string]{
			Path:      host,
			Env:       "WAVEFRONT_HOST",
			Argument:  host,
			Shorthand: "",
			Default:   "127.0.0.1",
			Usage:     "the host of the wavefront proxy",
			Value:     &handlerConfig.Host,
		},
		&sensu.PluginConfigOption[int]{
			Path:      port,
			Env:       "WAVEFRONT_METRICS_PORT",
			Argument:  port,
			Shorthand: "m",
			Default:   2878,
			Usage:     "the port of the wavefront proxy",
			Value:     &handlerConfig.MetricsPort,
		},
		&sensu.PluginConfigOption[int]{
			Path:      flush,
			Env:       "WAVEFRONT_FLUSH_INTERVAL_SECONDS",
			Argument:  flush,
			Shorthand: "f",
			Default:   1,
			Usage:     "the flush interval of the wavefront proxy (in seconds)",
			Value:     &handlerConfig.FlushIntervalSeconds,
		},
		&sensu.PluginConfigOption[string]{
			Path:      prefix,
			Env:       "WAVEFRONT_PREFIX",
			Argument:  prefix,
			Shorthand: "p",
			Usage:     "the string to be prepended to the metric name",
			Value:     &handlerConfig.Prefix,
		},
		&sensu.MapPluginConfigOption[string]{
			Path:      tags,
			Env:       "WAVEFRONT_TAGS",
			Argument:  tags,
			Shorthand: "t",
			Usage:     "the additional tags to merge with the metric tags",
			Value:     &handlerConfig.Tags,
		},
	}
)

func main() {
	handler := sensu.NewGoHandler(&handlerConfig.PluginConfig, opts, checkArgs, executeHandler)
	handler.Execute()
}

func checkArgs(event *corev2.Event) error {
	if !event.HasMetrics() {
		return fmt.Errorf("event does not contain metrics")
	}
	return nil
}

func executeHandler(event *corev2.Event) error {
	if len(event.Metrics.Points) == 0 {
		log.Println("event does not contain metric points")
		return nil
	}

	proxyCfg := &wavefront.ProxyConfiguration{
		Host:                 handlerConfig.Host,
		MetricsPort:          handlerConfig.MetricsPort,
		FlushIntervalSeconds: handlerConfig.FlushIntervalSeconds,
	}

	sender, err := wavefront.NewProxySender(proxyCfg)
	if err != nil {
		return err
	}

	defer sender.Close()

	for _, point := range event.Metrics.Points {
		tags := make(map[string]string)
		// merge tags if provided as config option
		if handlerConfig.Tags != nil {
			for k, v := range handlerConfig.Tags {
				tags[k] = v
			}
		}
		// overwrite tags with those from the original event
		for _, tag := range point.Tags {
			tags[tag.Name] = tag.Value
		}

		// prefix metric name if provided as config option
		name := point.Name
		if handlerConfig.Prefix != "" {
			name = fmt.Sprintf("%s.%s", handlerConfig.Prefix, name)
		}
		err := sender.SendMetric(name, point.Value, secTimestamp(point.Timestamp), event.Entity.Name, tags)
		if err != nil {
			log.Printf("error sending metric: %s", err)
		}
	}

	err = sender.Flush()
	log.Printf("sent %d metric points with %d failures", len(event.Metrics.Points), sender.GetFailureCount())
	return err
}

// msTimestamp auto-detection of metric point timestamp precision using a heuristic with a 250-ish year cutoff
func secTimestamp(ts int64) int64 {
	timestamp := ts
	switch ts := math.Log10(float64(timestamp)); {
	case ts < 10:
		// assume timestamp is seconds
	case ts < 13:
		// assume timestamp is milliseconds
		timestamp = (timestamp / 1e3)
	case ts < 16:
		// assume timestamp is microseconds
		timestamp = (timestamp / 1e6)
	default:
		// assume timestamp is nanoseconds
		timestamp = (timestamp / 1e9)
	}

	return timestamp
}
